{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats as ss\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_vandervej = pd.read_csv('snps_vandervej.tsv', sep=' ',\n",
    "                        names=['CHR', 'SNP', 'Position', 'EA(Freq)/NEA', 'Beta', 'SE', \n",
    "                               'p1', 'p2', 'p3', 'Candidate gene', 'Trait'])\n",
    "\n",
    "snps_ramirez_rec = pd.read_csv('snps_ramirez_rec.tsv', sep=' ')\n",
    "snps_ramirez_ex = pd.read_csv('snps_ramirez_ex.tsv', sep=' ')\n",
    "snps_ramirez_hr = pd.read_csv('snps_ramirez_hr.tsv', sep='\\t')\n",
    "snps_m2pam = pd.read_csv('snps_m2pam.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_literature = set(snps_vandervej['SNP']) | set(snps_ramirez_rec['SNP']) \\\n",
    "                  | set(snps_ramirez_ex['SNP']) | set(snps_m2pam['SNP'])\n",
    "snps_literature_df = pd.DataFrame(list(snps_literature))\n",
    "snps_literature_df.to_csv('snps_literature.tsv', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenos = glob.glob('hrr_phenotypes/phenotype*.tsv')\n",
    "pheno_dic = {}\n",
    "for pheno in phenos:\n",
    "    key = pheno.split('/')[-1].split('.')[-2]\n",
    "    pheno_dic[key] = pd.read_csv(pheno, sep='\\t')\n",
    "covariates = pd.read_csv('covariate_tables_fixed.csv', sep='\\t')\n",
    "pheno_set = set(covariates['FID'])\n",
    "for key in pheno_dic:\n",
    "    pheno_set = pheno_set & set(pheno_dic[key]['FID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_yes = pd.DataFrame(list(pheno_set), columns=['FID'])\n",
    "pheno_yes['IID'] = pheno_yes['FID'].copy(deep=True)\n",
    "pheno_no = pd.DataFrame(list(set(covariates['FID']) - set(pheno_yes['FID'])), columns=['FID'])\n",
    "pheno_no['IID'] = pheno_no['FID'].copy(deep=True)\n",
    "pheno_yes[['FID', 'IID']].to_csv('keep_ids_hrr.tsv', sep='\\t', index=False, header=False)\n",
    "pheno_no[['FID', 'IID']].to_csv('remove_ids_hrr.tsv', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariates = covariates.merge(pheno_yes, on='FID')\n",
    "covariates.to_csv('covariate_table_fixed_hrr.csv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Filter GWAS results\n",
    "# gwas_dirs = glob.glob('gwas_results/*')\n",
    "# gwases = {}\n",
    "# for gwas_dir in gwas_dirs:\n",
    "#     gwas = ''.join(gwas_dir.split('/')[-1].split('_')[2:])\n",
    "#     output = pd.read_csv(f'{gwas_dir}/bolt_output', sep='\\t', \n",
    "#                          usecols=['SNP', 'CHR', 'BP', 'BETA', 'ALLELE1', 'ALLELE0', 'P_BOLT_LMM'])\n",
    "#     imputed = pd.read_csv(f'{gwas_dir}/bolt_imputed_snps', sep='\\t', \n",
    "#                           usecols=['SNP', 'CHR', 'BP', 'BETA', 'ALLELE1', 'ALLELE0', 'P_BOLT_LMM'])\n",
    "#     gwases[gwas] = pd.concat([output[output['P_BOLT_LMM']<0.1], imputed[imputed['P_BOLT_LMM']<0.1]])\n",
    "#     gwases[gwas].to_csv(f'{gwas_dir}/filtered_snps.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Weight</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>MAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs2350782</td>\n",
       "      <td>0.137661</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.114009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs6962027</td>\n",
       "      <td>0.132023</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.456869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs73158705</td>\n",
       "      <td>0.146181</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.162847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs58394792</td>\n",
       "      <td>0.118127</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.163854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs28578872</td>\n",
       "      <td>-0.195589</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.433606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs7800170</td>\n",
       "      <td>-0.061350</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.492374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SNP    Weight A1 A2       MAF\n",
       "0   rs2350782  0.137661  T  C  0.114009\n",
       "1   rs6962027  0.132023  T  A  0.456869\n",
       "2  rs73158705  0.146181  A  G  0.162847\n",
       "3  rs58394792  0.118127  G  A  0.163854\n",
       "4  rs28578872 -0.195589  T  C  0.433606\n",
       "5   rs7800170 -0.061350  C  A  0.492374"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_m2pam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read GWAS results\n",
    "gwas_dirs = glob.glob('gwas_results/*')\n",
    "gwases = {}\n",
    "for gwas_dir in gwas_dirs:\n",
    "    gwas = ''.join(gwas_dir.split('/')[-1].split('_')[2:])\n",
    "    gwases[gwas] = pd.read_csv(f'{gwas_dir}/filtered_snps.tsv', sep='\\t')\n",
    "    idx = gwases[gwas]['SNP'].drop_duplicates().index\n",
    "    gwases[gwas] = gwases[gwas].loc[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Position</th>\n",
       "      <th>EA(Freq)/NEA</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>p3</th>\n",
       "      <th>Candidate gene</th>\n",
       "      <th>Trait</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs11589125</td>\n",
       "      <td>31894396</td>\n",
       "      <td>T(0.06)/C</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.013</td>\n",
       "      <td>6.6</td>\n",
       "      <td>×</td>\n",
       "      <td>10−09</td>\n",
       "      <td>SERINC2n,c</td>\n",
       "      <td>HRR50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs272564</td>\n",
       "      <td>45012273</td>\n",
       "      <td>A(0.71)/C</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.007</td>\n",
       "      <td>1.4</td>\n",
       "      <td>×</td>\n",
       "      <td>10−12</td>\n",
       "      <td>RNF220n,h</td>\n",
       "      <td>HRR50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs61765646</td>\n",
       "      <td>72723211</td>\n",
       "      <td>A(0.19)/T</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.008</td>\n",
       "      <td>1.1</td>\n",
       "      <td>×</td>\n",
       "      <td>10−13</td>\n",
       "      <td>NEGR1n</td>\n",
       "      <td>HRR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>rs1899492</td>\n",
       "      <td>60000304</td>\n",
       "      <td>T(0.47)/C</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.006</td>\n",
       "      <td>1.7</td>\n",
       "      <td>×</td>\n",
       "      <td>10−11</td>\n",
       "      <td>Gene_desert</td>\n",
       "      <td>HRR40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>rs17362588</td>\n",
       "      <td>179721046</td>\n",
       "      <td>G(0.92)/A</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.011</td>\n",
       "      <td>3.1</td>\n",
       "      <td>×</td>\n",
       "      <td>10−09</td>\n",
       "      <td>CCDC141n,c,TTNh</td>\n",
       "      <td>HRR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>rs35596070</td>\n",
       "      <td>179759692</td>\n",
       "      <td>C(0.86)/A</td>\n",
       "      <td>0.060</td>\n",
       "      <td>0.008</td>\n",
       "      <td>4.2</td>\n",
       "      <td>×</td>\n",
       "      <td>10−13</td>\n",
       "      <td>CCDC141n,c,TTNh</td>\n",
       "      <td>HRR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>rs73043051</td>\n",
       "      <td>18883863</td>\n",
       "      <td>C(0.22)/T</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.007</td>\n",
       "      <td>7.8</td>\n",
       "      <td>×</td>\n",
       "      <td>10−09</td>\n",
       "      <td>KCNH8n</td>\n",
       "      <td>HRR50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>rs34310778</td>\n",
       "      <td>74783408</td>\n",
       "      <td>C(0.43)/T</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.006</td>\n",
       "      <td>1.0</td>\n",
       "      <td>×</td>\n",
       "      <td>10−09</td>\n",
       "      <td>CNTN3n,e</td>\n",
       "      <td>HRR30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5</td>\n",
       "      <td>rs4836027</td>\n",
       "      <td>121866990</td>\n",
       "      <td>T(0.68)/C</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.006</td>\n",
       "      <td>1.7</td>\n",
       "      <td>×</td>\n",
       "      <td>10−15</td>\n",
       "      <td>SNCAIPn,PRDM6n,h</td>\n",
       "      <td>HRinc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5</td>\n",
       "      <td>rs151283</td>\n",
       "      <td>122446619</td>\n",
       "      <td>C(0.72)/A</td>\n",
       "      <td>0.042</td>\n",
       "      <td>0.007</td>\n",
       "      <td>1.6</td>\n",
       "      <td>×</td>\n",
       "      <td>10−10</td>\n",
       "      <td>PRDM6nh</td>\n",
       "      <td>HRR50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6</td>\n",
       "      <td>rs2224202</td>\n",
       "      <td>102053814</td>\n",
       "      <td>A(0.19)/G</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.007</td>\n",
       "      <td>5.8</td>\n",
       "      <td>×</td>\n",
       "      <td>10−09</td>\n",
       "      <td>GRIK2n,h</td>\n",
       "      <td>HRR20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7</td>\n",
       "      <td>rs2158712</td>\n",
       "      <td>26582733</td>\n",
       "      <td>A(0.52)/T</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.006</td>\n",
       "      <td>2.8</td>\n",
       "      <td>×</td>\n",
       "      <td>10−13</td>\n",
       "      <td>SKAPn,h</td>\n",
       "      <td>HRR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>7</td>\n",
       "      <td>rs180238</td>\n",
       "      <td>93550447</td>\n",
       "      <td>T(0.65)/C</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.006</td>\n",
       "      <td>2.2</td>\n",
       "      <td>×</td>\n",
       "      <td>10−12</td>\n",
       "      <td>GNG11n,GNGT1n,e,TFPI2n,e</td>\n",
       "      <td>HRR40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>7</td>\n",
       "      <td>rs3757868</td>\n",
       "      <td>100482720</td>\n",
       "      <td>G(0.82)/A</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.008</td>\n",
       "      <td>5.6</td>\n",
       "      <td>×</td>\n",
       "      <td>10−24</td>\n",
       "      <td>SRRTn,e,ACHEn,e,TRIP6e,C7orf43n,e,UFSP1n</td>\n",
       "      <td>HRR40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>7</td>\n",
       "      <td>rs1997571</td>\n",
       "      <td>116198621</td>\n",
       "      <td>A(0.58)/G</td>\n",
       "      <td>0.042</td>\n",
       "      <td>0.006</td>\n",
       "      <td>1.7</td>\n",
       "      <td>×</td>\n",
       "      <td>10−12</td>\n",
       "      <td>CAV1n,h,CAV2n,e,h</td>\n",
       "      <td>HRR20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>7</td>\n",
       "      <td>rs17168815</td>\n",
       "      <td>136624621</td>\n",
       "      <td>G(0.84)/T</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.008</td>\n",
       "      <td>1.1</td>\n",
       "      <td>×</td>\n",
       "      <td>10−14</td>\n",
       "      <td>CHRM2n</td>\n",
       "      <td>HRR50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>10</td>\n",
       "      <td>rs7072737</td>\n",
       "      <td>102556175</td>\n",
       "      <td>A(0.11)/G</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0.009</td>\n",
       "      <td>1.1</td>\n",
       "      <td>×</td>\n",
       "      <td>10−17</td>\n",
       "      <td>PAX2n</td>\n",
       "      <td>HRR40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>11</td>\n",
       "      <td>rs7130652</td>\n",
       "      <td>71984398</td>\n",
       "      <td>T(0.07)/G</td>\n",
       "      <td>0.076</td>\n",
       "      <td>0.011</td>\n",
       "      <td>3.4</td>\n",
       "      <td>×</td>\n",
       "      <td>10−11</td>\n",
       "      <td>CLPBn,h,INPPL1n,e</td>\n",
       "      <td>HRR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>12</td>\n",
       "      <td>rs4963772</td>\n",
       "      <td>24758480</td>\n",
       "      <td>A(0.15)/G</td>\n",
       "      <td>0.090</td>\n",
       "      <td>0.008</td>\n",
       "      <td>1.2</td>\n",
       "      <td>×</td>\n",
       "      <td>10−28</td>\n",
       "      <td>BCAT1n</td>\n",
       "      <td>HRR40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>12</td>\n",
       "      <td>rs6488162</td>\n",
       "      <td>33593127</td>\n",
       "      <td>C(0.58)/T</td>\n",
       "      <td>0.103</td>\n",
       "      <td>0.006</td>\n",
       "      <td>2.6</td>\n",
       "      <td>×</td>\n",
       "      <td>10−66</td>\n",
       "      <td>SYT10n,ALG10h</td>\n",
       "      <td>HRR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>12</td>\n",
       "      <td>rs61928421</td>\n",
       "      <td>116227249</td>\n",
       "      <td>C(0.93)/T</td>\n",
       "      <td>0.090</td>\n",
       "      <td>0.012</td>\n",
       "      <td>4.3</td>\n",
       "      <td>×</td>\n",
       "      <td>10−15</td>\n",
       "      <td>MED13Ln</td>\n",
       "      <td>HRR40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>14</td>\n",
       "      <td>rs17180489</td>\n",
       "      <td>72885471</td>\n",
       "      <td>C(0.14)/G</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.008</td>\n",
       "      <td>2.5</td>\n",
       "      <td>×</td>\n",
       "      <td>10−11</td>\n",
       "      <td>RGS6n,h</td>\n",
       "      <td>HRinc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>15</td>\n",
       "      <td>rs12906962</td>\n",
       "      <td>95312071</td>\n",
       "      <td>T(0.67)/C</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.006</td>\n",
       "      <td>2.7</td>\n",
       "      <td>×</td>\n",
       "      <td>10−14</td>\n",
       "      <td>MCTP2n</td>\n",
       "      <td>HRinc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>19</td>\n",
       "      <td>rs12974440</td>\n",
       "      <td>5894386</td>\n",
       "      <td>G(0.92)/A</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.011</td>\n",
       "      <td>2.4</td>\n",
       "      <td>×</td>\n",
       "      <td>10−10</td>\n",
       "      <td>FUT5n,NDUFA11n,c</td>\n",
       "      <td>HRR10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>19</td>\n",
       "      <td>rs12986417</td>\n",
       "      <td>30109533</td>\n",
       "      <td>G(0.65)/A</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.006</td>\n",
       "      <td>1.0</td>\n",
       "      <td>×</td>\n",
       "      <td>10−09</td>\n",
       "      <td>POP4n,C19orf12h</td>\n",
       "      <td>HRinc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR         SNP   Position EA(Freq)/NEA   Beta     SE   p1 p2     p3  \\\n",
       "0     1  rs11589125   31894396    T(0.06)/C  0.075  0.013  6.6  ×  10−09   \n",
       "1     1    rs272564   45012273    A(0.71)/C  0.046  0.007  1.4  ×  10−12   \n",
       "2     1  rs61765646   72723211    A(0.19)/T  0.056  0.008  1.1  ×  10−13   \n",
       "3     2   rs1899492   60000304    T(0.47)/C  0.040  0.006  1.7  ×  10−11   \n",
       "4     2  rs17362588  179721046    G(0.92)/A  0.062  0.011  3.1  ×  10−09   \n",
       "5     2  rs35596070  179759692    C(0.86)/A  0.060  0.008  4.2  ×  10−13   \n",
       "6     3  rs73043051   18883863    C(0.22)/T  0.041  0.007  7.8  ×  10−09   \n",
       "7     3  rs34310778   74783408    C(0.43)/T  0.036  0.006  1.0  ×  10−09   \n",
       "8     5   rs4836027  121866990    T(0.68)/C  0.050  0.006  1.7  ×  10−15   \n",
       "9     5    rs151283  122446619    C(0.72)/A  0.042  0.007  1.6  ×  10−10   \n",
       "10    6   rs2224202  102053814    A(0.19)/G  0.043  0.007  5.8  ×  10−09   \n",
       "11    7   rs2158712   26582733    A(0.52)/T  0.045  0.006  2.8  ×  10−13   \n",
       "12    7    rs180238   93550447    T(0.65)/C  0.043  0.006  2.2  ×  10−12   \n",
       "13    7   rs3757868  100482720    G(0.82)/A  0.077  0.008  5.6  ×  10−24   \n",
       "14    7   rs1997571  116198621    A(0.58)/G  0.042  0.006  1.7  ×  10−12   \n",
       "15    7  rs17168815  136624621    G(0.84)/T  0.062  0.008  1.1  ×  10−14   \n",
       "16   10   rs7072737  102556175    A(0.11)/G  0.079  0.009  1.1  ×  10−17   \n",
       "17   11   rs7130652   71984398    T(0.07)/G  0.076  0.011  3.4  ×  10−11   \n",
       "18   12   rs4963772   24758480    A(0.15)/G  0.090  0.008  1.2  ×  10−28   \n",
       "19   12   rs6488162   33593127    C(0.58)/T  0.103  0.006  2.6  ×  10−66   \n",
       "20   12  rs61928421  116227249    C(0.93)/T  0.090  0.012  4.3  ×  10−15   \n",
       "21   14  rs17180489   72885471    C(0.14)/G  0.055  0.008  2.5  ×  10−11   \n",
       "22   15  rs12906962   95312071    T(0.67)/C  0.048  0.006  2.7  ×  10−14   \n",
       "23   19  rs12974440    5894386    G(0.92)/A  0.067  0.011  2.4  ×  10−10   \n",
       "24   19  rs12986417   30109533    G(0.65)/A  0.037  0.006  1.0  ×  10−09   \n",
       "\n",
       "                              Candidate gene  Trait  \n",
       "0                                 SERINC2n,c  HRR50  \n",
       "1                                  RNF220n,h  HRR50  \n",
       "2                                     NEGR1n  HRR10  \n",
       "3                                Gene_desert  HRR40  \n",
       "4                            CCDC141n,c,TTNh  HRR10  \n",
       "5                            CCDC141n,c,TTNh  HRR10  \n",
       "6                                     KCNH8n  HRR50  \n",
       "7                                   CNTN3n,e  HRR30  \n",
       "8                           SNCAIPn,PRDM6n,h  HRinc  \n",
       "9                                    PRDM6nh  HRR50  \n",
       "10                                  GRIK2n,h  HRR20  \n",
       "11                                   SKAPn,h  HRR10  \n",
       "12                  GNG11n,GNGT1n,e,TFPI2n,e  HRR40  \n",
       "13  SRRTn,e,ACHEn,e,TRIP6e,C7orf43n,e,UFSP1n  HRR40  \n",
       "14                         CAV1n,h,CAV2n,e,h  HRR20  \n",
       "15                                    CHRM2n  HRR50  \n",
       "16                                     PAX2n  HRR40  \n",
       "17                         CLPBn,h,INPPL1n,e  HRR10  \n",
       "18                                    BCAT1n  HRR40  \n",
       "19                             SYT10n,ALG10h  HRR10  \n",
       "20                                   MED13Ln  HRR40  \n",
       "21                                   RGS6n,h  HRinc  \n",
       "22                                    MCTP2n  HRinc  \n",
       "23                          FUT5n,NDUFA11n,c  HRR10  \n",
       "24                           POP4n,C19orf12h  HRinc  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_vandervej"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwases_cross = {}\n",
    "for gwas in gwases:\n",
    "    gwases_cross[gwas] = {}\n",
    "    gwases_cross[gwas]['ramirez_rec'] = snps_ramirez_rec.merge(gwases[gwas], how='left', on='SNP', indicator=True)\n",
    "    gwases_cross[gwas]['ramirez_ex'] = snps_ramirez_rec.merge(gwases[gwas], how='left', on='SNP', indicator=True)\n",
    "    gwases_cross[gwas]['ramirez_hr'] = snps_ramirez_rec.merge(gwases[gwas], how='left', on='SNP', indicator=True)\n",
    "    gwases_cross[gwas]['vandervej'] = snps_vandervej.merge(gwases[gwas], how='left', on='SNP', indicator=True)\n",
    "    gwases_cross[gwas]['m2pam'] = snps_m2pam.merge(gwases[gwas], how='left', on='SNP', indicator=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "font = {'family' : 'normal',\n",
    "        'size'   : 16}\n",
    "\n",
    "matplotlib.rc('font', **font)\n",
    "\n",
    "f, ax = plt.subplots()\n",
    "xticklabels = []\n",
    "xticks = []\n",
    "gwases_keys = list(gwases.keys())\n",
    "gwases_keys = ['hrr', 'restinghr', 'deephrr', 'deephrrtransfer']\n",
    "gwases_labels = ['HRR', 'Resting HR', 'HRR from pretest', 'HRR from pretest and resting']\n",
    "for i, snp in enumerate(snps_ramirez_rec['SNP']):\n",
    "    xs = np.arange(len(gwases_keys))*3.0 + i * (len(gwases_keys)*3.0+10)\n",
    "    xticks.append(np.mean(xs))\n",
    "    xticklabels.append(snps_ramirez_rec.iloc[i]['Locus'])\n",
    "    arr = []\n",
    "    for j, gwas in enumerate(gwases_keys):\n",
    "        arr.append(gwases_cross[gwas]['ramirez_rec'].iloc[i]['P_BOLT_LMM'])\n",
    "        ax.bar(xs[j], -np.log10(arr[-1]), label=gwases_labels[j], width=3.0, edgecolor='black',\n",
    "               color=j*np.array([0.3, 0.3, 0.3]))\n",
    "    if i == 0 : \n",
    "        ax.legend(loc='upper left')\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/16.0), -np.log10(0.05/16.0)])\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/1e6), -np.log10(0.05/1e6)])\n",
    "ax.set_ylabel('$-log_{10}(p-value)$')\n",
    "f.set_size_inches(16, 9)\n",
    "ax.set_xticks(xticks)\n",
    "ax.set_xticklabels(xticklabels, rotation=45, ha='right')\n",
    "f.savefig('ramirez_rec.png', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_ramirez_rec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, ax = plt.subplots()\n",
    "xticklabels = []\n",
    "xticks = []\n",
    "gwases_keys = list(gwases.keys())\n",
    "gwases_keys = ['hrr', 'restinghr', 'deephrr', 'deephrrtransfer']\n",
    "for i, snp in enumerate(snps_ramirez_ex['SNP']):\n",
    "    xs = np.arange(len(gwases_keys))*3.0 + i * (len(gwases_keys)*3.0+10)\n",
    "    xticks.append(np.mean(xs))\n",
    "    xticklabels.append(snps_ramirez_ex.iloc[i]['Locus'])\n",
    "    arr = []\n",
    "    for j, gwas in enumerate(gwases_keys):\n",
    "        arr.append(gwases_cross[gwas]['ramirez_ex'].iloc[i]['P_BOLT_LMM'])\n",
    "        ax.bar(xs[j], -np.log10(arr[-1]), label=gwases_labels[j], width=3.0, edgecolor='black',\n",
    "               color=j*np.array([0.3, 0.3, 0.3]))\n",
    "    if i == 0 : \n",
    "        ax.legend(loc='upper left')\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/16.0), -np.log10(0.05/16.0)])\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/1e6), -np.log10(0.05/1e6)])\n",
    "f.set_size_inches(16, 9)\n",
    "ax.set_xticks(xticks)\n",
    "ax.set_xticklabels(xticklabels, rotation=45, ha='right')\n",
    "f.savefig('ramirez_ex.png', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_ramirez_ex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, ax = plt.subplots()\n",
    "xticklabels = []\n",
    "xticks = []\n",
    "gwases_keys = list(gwases.keys())\n",
    "gwases_keys = ['hrr', 'restinghr', 'deephrr', 'deephrrtransfer']\n",
    "for i, snp in enumerate(snps_vandervej['SNP']):\n",
    "    xs = np.arange(len(gwases_keys))*8.0 + i * (len(gwases_keys)*8.0+10)\n",
    "    xticks.append(np.mean(xs))\n",
    "    xticklabels.append(snps_vandervej.iloc[i]['Candidate gene'].split(',')[0])\n",
    "    arr = []\n",
    "    for j, gwas in enumerate(gwases_keys):\n",
    "        arr.append(gwases_cross[gwas]['vandervej'].iloc[i]['P_BOLT_LMM'])\n",
    "        ax.bar(xs[j], -np.log10(arr[-1]), label=gwases_labels[j], width=8, edgecolor='black',\n",
    "               color=j*np.array([0.3, 0.3, 0.3]))\n",
    "    if i == 0 : \n",
    "        ax.legend()\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/23.0), -np.log10(0.05/23.0)])\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/1e6), -np.log10(0.05/1e6)])\n",
    "f.set_size_inches(16, 9)\n",
    "ax.set_xticks(xticks)\n",
    "ax.set_xticklabels(xticklabels, rotation=45, ha='right')\n",
    "f.savefig('verweij.png', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_vandervej"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwases.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, ax = plt.subplots()\n",
    "xticklabels = []\n",
    "xticks = []\n",
    "gwases_keys = list(gwases.keys())\n",
    "gwases_keys = ['hrr', 'restinghr', 'deephrr', 'deephrrtransfer']\n",
    "for i, snp in enumerate(snps_m2pam['SNP']):\n",
    "    xs = np.arange(len(gwases_keys))*3.0 + i * (len(gwases_keys)*3.0+10)\n",
    "    xticks.append(np.mean(xs))\n",
    "    xticklabels.append(snps_m2pam.iloc[i]['SNP'])\n",
    "    arr = []\n",
    "    for j, gwas in enumerate(gwases_keys):\n",
    "        arr.append(gwases_cross[gwas]['m2pam'].iloc[i]['P_BOLT_LMM'])\n",
    "        ax.bar(xs[j], -np.log10(arr[-1]), label=gwases_labels[j], width=3, edgecolor='black',\n",
    "               color=j*np.array([0.3, 0.3, 0.3]))\n",
    "    if i == 0 : \n",
    "        ax.legend()\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/6.0), -np.log10(0.05/6.0)])\n",
    "ax.plot([0, np.max(xs)], [-np.log10(0.05/1e6), -np.log10(0.05/1e6)])\n",
    "f.set_size_inches(16, 9)\n",
    "ax.set_xticks(xticks)\n",
    "ax.set_xticklabels(xticklabels, rotation=45, ha='right')\n",
    "f.savefig('m2pam.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwases['hrr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_gwas(df_tmp, fig, ax):\n",
    "    df = df_tmp.sort_values(by=['CHR', 'BP'])\n",
    "    df['LOGP'] = -np.log10(df['P_BOLT_LMM'])\n",
    "    df.CHR = df.CHR.astype('category')\n",
    "    df['ind'] = range(len(df))\n",
    "    df_grouped = df.groupby(('CHR'))\n",
    "       \n",
    "    colors = ['black','silver']\n",
    "    x_labels = []\n",
    "    x_labels_pos = []\n",
    "\n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        group.plot(kind='scatter', x='ind', y='LOGP',color=colors[num % len(colors)], ax=ax, s=5)\n",
    "        x_labels.append(name)\n",
    "        x_labels_pos.append((group['ind'].iloc[-1] - (group['ind'].iloc[-1] - group['ind'].iloc[0])/2))   \n",
    "        \n",
    "    ax.plot([0.0, group['ind'].iloc[-1]], [7.3, 7.3])\n",
    "    \n",
    "    ax.set_xticks(x_labels_pos)\n",
    "    ax.set_xticklabels(x_labels)\n",
    "    ax.set_xlim([0, len(df)])\n",
    "    ax.set_ylim([1, 20])\n",
    "    ax.set_xlabel('Chromosome')\n",
    "    ax.set_ylabel('$-\\log_{10}(p-value)$')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for gwas in gwases:\n",
    "    f, ax = plt.subplots()\n",
    "    f.set_size_inches(22, 6)\n",
    "    plot_gwas(gwases[gwas], f, ax)\n",
    "    f.savefig(f'manhattan_{gwas}.png', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from assocplots.qqplot import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter GWAS results\n",
    "gwas_dirs = glob.glob('gwas_results/*')\n",
    "gwases_outputs = {}\n",
    "for gwas_dir in gwas_dirs:\n",
    "    gwas = ''.join(gwas_dir.split('/')[-1].split('_')[2:])\n",
    "    output = pd.read_csv(f'{gwas_dir}/bolt_output', sep='\\t', \n",
    "                         usecols=['SNP', 'CHR', 'BP', 'BETA', 'ALLELE1', 'ALLELE0', 'P_BOLT_LMM'])\n",
    "    # imputed = pd.read_csv(f'{gwas_dir}/bolt_imputed_snps', sep='\\t', \n",
    "    #                      usecols=['SNP', 'CHR', 'BP', 'BETA', 'ALLELE1', 'ALLELE0', 'P_BOLT_LMM'])\n",
    "    gwases_outputs[gwas] = output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for gwas in gwases_outputs:\n",
    "    f, ax = plt.subplots()\n",
    "    f.set_size_inches(6, 6)\n",
    "    qqplot([gwases_outputs[gwas]['P_BOLT_LMM']], \n",
    "           [gwas], \n",
    "           color=['k'], \n",
    "           fill_dens=[0.2], \n",
    "           error_type='theoretical', \n",
    "           distribution='beta',\n",
    "           title='')\n",
    "    f.savefig(f'qq_{gwas}.png', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, ax = plt.subplots()\n",
    "f.set_size_inches(5, 5)\n",
    "plt.sca(ax)\n",
    "qqplot([gwas_dic['HRR']['P_BOLT_LMM']], \n",
    "       ['HRR (UKBB)'], \n",
    "       color=['k'], \n",
    "       fill_dens=[0.2], \n",
    "       error_type='theoretical', \n",
    "       distribution='beta',\n",
    "       title='HRR (UKBB)')\n",
    "f.savefig('HRR_qq.png')\n",
    "\n",
    "# plt.sca(ax[1])\n",
    "# qqplot([df_deep_hrr['P_BOLT_LMM']], \n",
    "#        ['HRR from pretest'], \n",
    "#        color=['k'], \n",
    "#        fill_dens=[0.2], \n",
    "#        error_type='theoretical', \n",
    "#        distribution='beta',\n",
    "#        title='HRR from pretest')\n",
    "\n",
    "# plt.sca(ax[2])\n",
    "# qqplot([df_deep_transfer['P_BOLT_LMM']], \n",
    "#        ['HRR from pretest and resting ECGs'], \n",
    "#        color=['k'], \n",
    "#        fill_dens=[0.2], \n",
    "#        error_type='theoretical', \n",
    "#        distribution='beta',\n",
    "#        title='HRR from pretest and resting ECGs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
